Added side chain feature to poly synth

samhincks opened the pull request:
A side chain feature has been added to the poly-synth. 
- There is a new class src/noisegate-side-chain which is identical to the audio-worklet/src/noisegate-script-processor.js except it always processes stereo input, builds an envelope on the first channel, and mutes the second channel when the first channel exceeds the threshold, returning the contents of the first channel.
- index, poly-synth-demo, and poly-synth.js have been modified to accomodate side chaining. While there is a lot of new code, it is similar in form to that which already exists.
- I changed the presets simply so that the demo can be tested, but will change it back to @hoch's preset before merging. 

The demo can be found [here](http://rawgit.com/GoogleChrome/web-audio-samples/7192edcd88788d08a8dc2602179f9b2c6c785c8e/poly-synth/index.html).

rtoy commented:
Update the title of this PR. You're not adding a side chain to the noise gate code. You're adding a (new) noise gate to the synth.

rtoy commented:
Notes on the demo:  

- When the side chain is started, why can't I move the playback rate slider, but I can for the others?
- Why is the reverb response fixed to matrix-reverb1 now?
- Are you going to have more than one beat file for the side chain?
- Should the beat always be audible? Just asking because I don't know what other synths do with that.
- Introductory text doesn't say anything about the side chain.
- Is "Side chain" really the right name? Not Noise gate?
- If the side chain is just a simple beat, should the "playback rate" slider really be beats per minute or somehing?

rtoy commented on poly-synth/src/noisegate-side-chain.js:
```diff
@@ -0,0 +1,212 @@
+        timeConstant, this.context_.sampleRate);
+
+    this.noiseGateKernel_ = this.context_.createScriptProcessor(
+        bufferSize, 2, 1);
```
I think this means the audio input to the noise gate is mono?  I was expecting at least stereo.

samhincks replied:
The input to the noise gate has 2 channels and the output has one channel.

hoch replied:
Well, this is a bit different from what we discussed. Perhaps this is fine for the use case because the synth is already mono channel. However, if you're building a more generic Noisegate it your kernel should have 4 channel input and 2 channel output.

rtoy commented on poly-synth/src/noisegate-side-chain.js:
```diff
@@ -0,0 +1,212 @@
+    this.noiseGateKernel_.onaudioprocess = this.onaudioprocess_.bind(this);
+
+    // The noise gate is connected to and from by dummy input and output nodes.
+    this.input = new GainNode(this.context_);
```
Shouldn't the noise gate have two inputs? One for the audio signal and one for the side chain signal?

samhincks replied:
@hoch and I discussed this, and decided the easiest solution would be to let the input have two channels and use one channel for the envelope and modify the contents of the second channel. 

I'm not sure how an implementation would work for the two node solution you describe.

hoch replied:
That's NOT what we discussed. I suggested to create a subgraph with splitter and merger so the kernel can process 4 channels. Perhaps you can sum the sidechain input to mono because it is being summed down in the script processor node anyway.

rtoy commented on poly-synth/src/noisegate-side-chain.js:
```diff
@@ -0,0 +1,212 @@
+   */
+  onaudioprocess_(event) {
+    let inputBuffer = event.inputBuffer;
+    let envelopeChannel = inputBuffer.getChannelData(0);
```
What is the `envelopeChannel`?  The name seems to imply that output of the envelope follower, but it seems this is the side chain signal that feeds the envelope follower.

samhincks replied:
I will rename this to sideChainChannel.

samhincks commented on poly-synth/src/noisegate-side-chain.js:
```diff
@@ -0,0 +1,212 @@
+  onaudioprocess_(event) {
+    let inputBuffer = event.inputBuffer;
+    let envelopeChannel = inputBuffer.getChannelData(0);
+    let sideChainChannel = inputBuffer.getChannelData(1);
```
Ah, I looked up and you're right that I got the name backwards. The channel you build the envelope on is referred to as the side chain. 

All the synthesizer audio is in one channel `inputBuffer.getChannelData(1);|` and all the data for the drum sample is in `inputBuffer.getChannelData(0);|` . So the input has 2 channels but they are completely different. This makes it a stereo input, right?

hoch replied:
If you implement this correctly, this is what you'll get:

`inputBuffer.getChannelData(0)`: actual audio input ch 1
`inputBuffer.getChannelData(1)`: actual audio input ch 2
`inputBuffer.getChannelData(2)`: side chain input ch 1
`inputBuffer.getChannelData(3)`: side chain input ch 2

rtoy replied:
I agree with hoch@. Not sure if that's needed at this point.  If you were doing this for the actual noise gate, this would be a requirement.

samhincks commented on poly-synth/src/poly-synth.js:
```diff
@@ -103,15 +115,47 @@ class PolySynth {
-
-    this.voiceOutput_.connect(bitcrusherInputNode);
+    this.voiceOutput_.connect(noisegateInputNode);
+    noisegateOutputNode.connect(bitcrusherInputNode);
```
My reasoning was that it doesn't really matter since the noise gate isn't going to construct an envelope the synthesizer data anyway. I put the noise gate first since, in theory one could optimize the bit crusher so that it didn't do any work when its input was all 0's.

hoch replied:
Doesn't matter. It's a design choice. However, you have to clearly indicate the flow of the signal somehow. Otherwise, user will have no clue at all. How about a diagram in SVG?

samhincks commented on poly-synth/src/poly-synth.js:
```diff
@@ -103,15 +115,47 @@ class PolySynth {
+        this.activeNoisegateRoute.connect(noisegateOutputNode));
+
+    // If |this.drumVolume| is positive, the drum sample is audible but bypasses
+    // the other synthesizer effects.
```
Exactly. I want the user to be able to control the volume of the drum sample. Sometimes you want to hear the two interact and other times you simply want the rhythm information from the beat sample.

rtoy replied:
Then it's not really part of the side chain, then, I think.  It's really a gain knob for the beat track. Not sure how to address this in the app.

samhincks commented on poly-synth/src/poly-synth-demo.js:
```diff
@@ -224,6 +248,109 @@ class PolySynthDemo {
+      // The change is scheduled slightly into the future to avoid glitching.
+      if (event.target.textContent === 'Start') {
+        event.target.textContent = 'Stop';
+        this.drumSamplePlaybackRateSlider_.disable();
```
My mistake. The next version will of course support this feature.

samhincks commented:
> When the side chain is started, why can't I move the playback rate slider, but I can for the others?
I will change this feature in the next push. 

> Why is the reverb response fixed to matrix-reverb1 now?
@hoch found a nice preset, using this impulse response.

> Are you going to have more than one beat file for the side chain?
Yes. This is currently in the [demo](http://rawgit.com/GoogleChrome/web-audio-samples/7192edcd88788d08a8dc2602179f9b2c6c785c8e/poly-synth/index.html) in the selector under side chain.

> Should the beat always be audible? Just asking because I don't know what other synths do with that.
The user can control the audible volume of the beat, which I think is the right decision so that the beat can be used as a rhythm settor and so that you can hear the interaction between beat and synth. The volume of the beat that is fed to the noise gate is fixed.

> Is "Side chain" really the right name? Not Noise gate?
I'm not sure on this one. I think you would say that you "side chain a beat to a synthesizer using a noise gate". What does @hoch think?

> If the side chain is just a simple beat, should the "playback rate" slider really be beats per minute or somehing?
There are more complex beats in there.

samhincks commented:
In the latest [commit](https://github.com/GoogleChrome/web-audio-samples/pull/116/commits/68b2414e1f1cc13466125f41d70ed93ce9eed75b), 
- modifications based on review

The demo is [here](http://rawgit.com/GoogleChrome/web-audio-samples/68b2414e1f1cc13466125f41d70ed93ce9eed75b/poly-synth/index.html)

rtoy commented on poly-synth/index.html:
```diff
@@ -66,6 +71,11 @@
         <div id="filter-envelope" class="controls"><h2>
           Filter Envelope</h2>
         </div>
+        <div id="sideChain" class="controls">
```
Shouldn't this be noisegate instead of sideChain?

rtoy commented on poly-synth/src/noisegate-side-chain.js:
```diff
@@ -0,0 +1,215 @@
+
+  /**
+   * Compute an envelope follower for the signal.
+   * @param {Float32Array} channelData Input channel data.
```
Not really channel data anymore, right?

rtoy commented on poly-synth/src/poly-synth-demo.js:
```diff
@@ -224,6 +248,108 @@ class PolySynthDemo {
+          default: this.polySynth_.drumVolume
+        });
+
+    this.sideChainThresholdSlider_ = new ParamController(
```
Not really sideChain anymore.

rtoy commented on poly-synth/src/poly-synth.js:
```diff
@@ -103,15 +115,50 @@ class PolySynth {
     reverbOutputNode.connect(this.output);
   }
 
+  createNoisegateSideChainGraph_(inputNode) {
```
I really think there should be two inputs, one for the signal and one for the side chain control, but too late for that now, I think.

rtoy commented on poly-synth/src/poly-synth.js:
```diff
@@ -103,15 +115,50 @@ class PolySynth {
+
+    // The noise gate will build an envelope on the output of the drum sample
+    // (in channel 0) but the noise gate will modify the output of the
+    // synthesizer (channel 1).
```
This no longer matches the actual implementation, I think.

rtoy commented on poly-synth/src/poly-synth.js:
```diff
@@ -249,6 +296,76 @@ class PolySynth {
 
   /**
+   * Set the AudioBuffer of the drum sample used for side chaining.
+   * @param {AudioBuffer} impulseResponseBuffer The new impulse response buffer.
```
Not an "impulseResponseBuffer".

rtoy commented on poly-synth/src/poly-synth-demo.js:
```diff
@@ -224,6 +248,108 @@ class PolySynthDemo {
+      option.textContent = abbreviatedUrl;
+      selector.appendChild(option);
+    }
+
```
You do exactly the same thing for the reverb.  It would be nice to factor this common code out.

rtoy commented on poly-synth/src/poly-synth-demo.js:
```diff
@@ -61,17 +70,32 @@ class PolySynthDemo {
+    for (let index in drumSamples) {
+      drumSampleBuffers[index] = await this.loadSound(drumSamples[index]);
+    }
+    return drumSampleBuffers;
```
Don't you do exactly the same for the reverb?  It would be nice to factor out this common code.

rtoy commented on poly-synth/src/noisegate-side-chain.js:
```diff
@@ -0,0 +1,215 @@
+
+/**
+ * @class NoiseGateSideChain
+ * @description This script-processor-wrapper class is replica of
```
Not really a replica, right?

rtoy commented on poly-synth/index.html:
```diff
@@ -55,7 +56,11 @@
-    reverb, and delay effects.
+    reverb, and delay effects. If noise gate side chaining is active, then
+    the output of the synthesizer will ramp towards silence when the level of
+    the drum sample exceeds the specified threshold. The beat and its volume is
```
Isn't this backwards?  If the drum sample exceeds the threshold, the synthesizer ramps up, not down to silence.

samhincks replied:
No, the drum interrupts the sustained note. When the volume of the drum exceeds the threshold, the synthesizer will map towards silence because we subtract the weight from 1 `outputChannel0[j] = (1 - weights[j]) * signalChannel0[j];`

rtoy commented:
A bug in playing the demo:  Load the page.  Select d2-60 for the drum sample.  Start noise gate.  Press any key. I can't hear the note, even if I set the drum volume to 0.  Stop the noise gate.  Press a key. Note is heard.

Perhaps not a bug, but I was kind of expecting to be able to switch drum samples at any time, even if the noise gate is on. I can only change if the gate is stopped. I can swap impulse responses at any time with the reverb.

Also, I loaded up the webaudio inspector.  There seems to be lots of gain nodes after the convolver.  Some seem unnecessary, but I didn't trace the nodes back to the code, so I'm not sure.

samhincks commented on poly-synth/index.html:
```diff
@@ -56,9 +56,11 @@
-    exceeds the specified threshold.
+    reverb, and delay effects. If noise gate side chaining is active, then
+    the output of the synthesizer will ramp towards silence when the level of
+    the drum sample exceeds the specified threshold. The beat and its volume is
```
It's backwards with respect to the original noise gate. However, we modify the noise gate side chain so that it behaves this way. 
`      outputChannel0[j] = (1 - weights[j]) * signalChannel0[j];
`

samhincks commented on poly-synth/index.html:
```diff
@@ -56,9 +56,11 @@
+    the output of the synthesizer will ramp towards silence when the level of
+    the drum sample exceeds the specified threshold. The beat and its volume is
+    conceptually separate from the synthesizer and changing synthesizer effects
+    does not modify its sound.
```
The beat. I have changed the sentence.

hoch commented on poly-synth/src/poly-synth-demo.js:
```diff
@@ -40,12 +40,21 @@ class PolySynthDemo {
+   * @param {String} identifiers.drumSelectorId The ID of the <select>
+   *                                            element for drum samples.
    */
   constructor(context, identifiers) {
```
`identifiers` does not any value at all. How about `elementIds`?

hoch commented on poly-synth/src/poly-synth-demo.js:
```diff
@@ -362,6 +488,22 @@ class PolySynthDemo {
   }
+  
+  getDrumSampleUrls_() {
+    let drumSampleUrls =
```
Again, this creates an array every time the method called. You should have this in a global data storage for this class.

hoch commented on poly-synth/src/noisegate-side-chain.js:
```diff
@@ -0,0 +1,215 @@
+    let outputChannel0 = event.outputBuffer.getChannelData(0);
+    let outputChannel1 = event.outputBuffer.getChannelData(1);
+    
+    let envelope = this.detectLevel_(sideChainChannel);
```
So I guess this always processes the audio with sidechain?

samhincks replied:
Yes, thats the intention.

hoch commented on poly-synth/src/poly-synth-demo.js:
```diff
@@ -363,6 +467,34 @@ class PolySynthDemo {
+  }
+  
+  getDrumSampleUrls_() {
+    let drumSampleUrls =
```
Please have a global storage for this list.

samhincks commented:
Latest commit is [here](https://github.com/GoogleChrome/web-audio-samples/pull/116/commits/cb205c37af42d1b0e29f6803d6f674c2645ecd16) and latest demo is [here](http://rawgit.com/GoogleChrome/web-audio-samples/cb205c37af42d1b0e29f6803d6f674c2645ecd16/poly-synth/index.html). You can now switch between drum samples at any time.

> A bug in playing the demo: Load the page. Select d2-60 for the drum sample. Start noise gate. Press any key. I can't hear the note, even if I set the drum volume to 0. Stop the noise gate. Press a key. Note is heard.

This isn't a bug since the drum volume does not affect the noise gate computation. The gain node of the drum connects directly to the output of the synth.

hoch commented:
An example of diagram would be: http://hoch.github.io/WAAX/creating-waax-plugin/

rtoy commented:
On Fri, Aug 18, 2017 at 8:42 AM, samhincks <notifications@github.com> wrote:

> Latest commit is here
> <https://github.com/GoogleChrome/web-audio-samples/pull/116/commits/cb205c37af42d1b0e29f6803d6f674c2645ecd16>
> and latest demo is here
> <http://rawgit.com/GoogleChrome/web-audio-samples/cb205c37af42d1b0e29f6803d6f674c2645ecd16/poly-synth/index.html>.
> You can now switch between drum samples at any time.
>
> A bug in playing the demo: Load the page. Select d2-60 for the drum
> sample. Start noise gate. Press any key. I can't hear the note, even if I
> set the drum volume to 0. Stop the noise gate. Press a key. Note is heard.
>
> This isn't a bug since the drum volume does not affect the noise gate
> computation. The gain node of the drum connects directly to the output of
> the synth.
>

​You misunderstand. I hear NOTHING. No note.  NOTHING. Any key press
produces NOTHING.​


> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleChrome/web-audio-samples/pull/116#issuecomment-323388596>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAofPB3Oetk3SXYLuokP1svhj51N4ucjks5sZbDygaJpZM4O6lXn>
> .
>



-- 
Ray

rtoy commented:
Sorry, you're right.  The beat sample selected always has something
playing, so the gate is closed.

On Fri, Aug 18, 2017 at 9:07 AM, Raymond Toy <toy.raymond@gmail.com> wrote:

>
>
> On Fri, Aug 18, 2017 at 8:42 AM, samhincks <notifications@github.com>
> wrote:
>
>> Latest commit is here
>> <https://github.com/GoogleChrome/web-audio-samples/pull/116/commits/cb205c37af42d1b0e29f6803d6f674c2645ecd16>
>> and latest demo is here
>> <http://rawgit.com/GoogleChrome/web-audio-samples/cb205c37af42d1b0e29f6803d6f674c2645ecd16/poly-synth/index.html>.
>> You can now switch between drum samples at any time.
>>
>> A bug in playing the demo: Load the page. Select d2-60 for the drum
>> sample. Start noise gate. Press any key. I can't hear the note, even if I
>> set the drum volume to 0. Stop the noise gate. Press a key. Note is heard.
>>
>> This isn't a bug since the drum volume does not affect the noise gate
>> computation. The gain node of the drum connects directly to the output of
>> the synth.
>>
>
> ​You misunderstand. I hear NOTHING. No note.  NOTHING. Any key press
> produces NOTHING.​
>
>
>> —
>> You are receiving this because your review was requested.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/GoogleChrome/web-audio-samples/pull/116#issuecomment-323388596>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAofPB3Oetk3SXYLuokP1svhj51N4ucjks5sZbDygaJpZM4O6lXn>
>> .
>>
>
>
>
> --
> Ray
>



-- 
Ray

samhincks commented:
In the last [commit](https://github.com/GoogleChrome/web-audio-samples/pull/116/commits/c66cb73c73309ca225db121d163b691f5e4ad8f9), I changed presets back to @hoch, added a flow chart and made minor modifications based on review. The demo is [here](http://rawgit.com/GoogleChrome/web-audio-samples/c66cb73c73309ca225db121d163b691f5e4ad8f9/poly-synth/index.html).

rtoy commented on poly-synth/src/noisegate-side-chain.js:
```diff
@@ -0,0 +1,211 @@
+        const weight = this.previousWeight_ - attackLossPerStep;
+        this.weights_[i] = Math.max(weight, 0);
+      }
+      else {
```
Nit: '} else {'

samhincks replied:
Nice catch :)
